### PR TITLE
Rover: add support for custom motor configurations 

### PR DESCRIPTION
**Pull request by Ammarf**
This PR adds support for custom motor configurations for Rover.

Similar to copter, sub, etc.., it uses a matrix with 3 factors, throttle, steering and lateral (perhaps forward, yaw and lateral is more suitable?), does the mixing and sends the output to each motor. This should allow the use of boats with side thrusters, surface vehicle with sub-like configs and rovers like the omni-rover. This PR also replaces the current omni-rover code as it works just well (it also fixes a bug in the omni-rover code I found a while ago relating to the rover's throttle), and it adds more functionality for somewhat the same amount of code. 

I tested it on the omni-rover ([video](https://youtu.be/GqyqBl8X9GQ)) and a regular rover, and they both seem to work well. Once you add the matrix for the custom frame, you can test it by setting the FRAME_TYPE parameter to 2, disabling all servo functions and rebooting. 

My only point of concern is that for frames that use normal throttle and steering control in addition to, say side thursters, they wont be using the usual functions like output_throttle(), so perhaps any future changes to how throttle and steering work in rover won't apply for these custom frames.

### CODE REVIEW

File: APMrover2/AP_MotorsUGV.cpp [hunk 1]
```diff
@@ -158,10 +163,86 @@ void AP_MotorsUGV::setup_servo_output()
...
+    switch (rover.g2.frame_type) {
+
+    //   FRAME TYPE NAME,            MOTOR NUMBER,           THROTTLE FACTOR,            STEERING FACTOR,            LATERAL FACTOR
+
+    case FRAME_TYPE_UNDEFINED:
+        break;
+
+    case FRAME_TYPE_OMNI:
+        _config_num_of_motors = 3;
+        add_config_motor(                 0,                       1.0f,                      1.0f,                       -1.0f);
```
Comment (khancyr) [inline APMrover2/AP_MotorsUGV.cpp]: I am not fan of those spaces ... Let's wait other opinion

File: APMrover2/AP_MotorsUGV.cpp [hunk 2]
```diff
@@ -591,6 +608,63 @@ void AP_MotorsUGV::output_skid_steering(bool armed, float steering, float thrott
...
+        // constrain steering
+        steering = constrain_float(steering, -4500.0f, 4500.0f);
+
+        // reverse steering when throttle is negative
+        if (is_negative(throttle)) {
+            steering *= -1.0f;
+        }
+
+        // scale throttle, steering and lateral inputs to -1 to 1
+        float scaled_throttle = throttle / 100.0f;
```
Comment (khancyr) [inline APMrover2/AP_MotorsUGV.cpp]: const those 3

File: APMrover2/Rover.h [hunk 3]
```diff
@@ -125,16 +125,17 @@ class Rover : public AP_HAL::HAL::Callbacks {
     void setup(void) override;
     void loop(void) override;
 
+    // all settable parameters
+    Parameters g;
+    ParametersG2 g2;
```
Comment (khancyr) [inline APMrover2/Rover.h]: why this change ?

File: APMrover2/AP_MotorsUGV.cpp [hunk 4]
```diff
@@ -591,6 +608,63 @@ void AP_MotorsUGV::output_skid_steering(bool armed, float steering, float thrott
...
+        float scaled_steering = steering / 4500.0f;
+        float scaled_lateral = lateral / 100.0f;
+
+        float thr_str_ltr_out[max_num_motors];
```
Comment (khancyr) [inline APMrover2/AP_MotorsUGV.cpp]: those are only use in the for loop, so no need of a table ... just reduce the scope

File: APMrover2/AP_MotorsUGV.cpp [hunk 5]
```diff
@@ -591,6 +608,63 @@ void AP_MotorsUGV::output_skid_steering(bool armed, float steering, float thrott
...
+
+        for (uint8_t i=0; i<max_num_motors; i++) {
+            if (_motor_enabled[i]) {
+                thr_str_ltr_out[i] = (scaled_throttle *_throttle_factor[i]) +
+                                     (scaled_steering *_steering_factor[i]) +
+                                     (scaled_lateral *_lateral_factor[i]);
+                if (fabsf(thr_str_ltr_out[i]) > thr_str_ltr_max) {
+                    thr_str_ltr_max = fabsf(thr_str_ltr_out[i]);
+                }
+                _output_vectored[i] = (thr_str_ltr_out[i] / thr_str_ltr_max);
```
Comment (khancyr) [inline APMrover2/AP_MotorsUGV.cpp]: this table isn't use in other place, so it isn't needed

File: APMrover2/AP_MotorsUGV.h [hunk 6]
```diff
@@ -10,6 +10,8 @@ class AP_MotorsUGV {
     // Constructor
     AP_MotorsUGV(AP_ServoRelayEvents &relayEvents);
 
+#define max_num_motors 6
```
Comment (khancyr) [inline APMrover2/AP_MotorsUGV.h]: static const int8_t UGV_MAX_NUM_MOTORS = 6;

**Comment by khancyr** [comment]
Hello, 
Thanks you for submitting that ! It is really interesting. I put some comments, I will do another pass later.

File: APMrover2/Rover.h (hunk 3 above)
Comment (Ammarf) [reply APMrover2/Rover.h]: If they're private I can't access frame_type parameter in AP_MotorsUGV.cpp.

File: APMrover2/AP_MotorsUGV.cpp [hunk 7]
```diff
@@ -115,6 +115,11 @@ void AP_MotorsUGV::init()
 
     // set safety output
     setup_safety_output();
+
+    // setup motors for custom configs
+    if (!FRAME_TYPE_UNDEFINED) {
```
Comment (khancyr) [inline APMrover2/AP_MotorsUGV.cpp]: rover.g2.frame_type != FRAME_TYPE_UNDEFINED

File: APMrover2/AP_MotorsUGV.cpp [hunk 8]
```diff
@@ -591,6 +608,62 @@ void AP_MotorsUGV::output_skid_steering(bool armed, float steering, float thrott
     output_throttle(SRV_Channel::k_throttleRight, 100.0f * motor_right);
 }
 
+// output for custom configurations
+void AP_MotorsUGV::output_custom_config(bool armed, float steering, float throttle, float lateral)
+{
+    // exit immediately if the frame type is set to UNDEFINED
+    if (FRAME_TYPE_UNDEFINED) {
```
Comment (khancyr) [inline APMrover2/AP_MotorsUGV.cpp]: rover.g2.frame_type == FRAME_TYPE_UNDEFINED

File: APMrover2/Rover.h (hunk 3 above)
Comment (khancyr) [reply APMrover2/Rover.h]: just make a getter : get_frame_type on rover class.

**Comment by Ammarf** [comment]
I added the getter and retested the code, works well.

File: APMrover2/AP_MotorsUGV.h [hunk 9]
```diff
@@ -138,5 +156,13 @@ class AP_MotorsUGV {
...
-    float   _lateral;  // requested lateral input as a value from -4500 to +4500
+    float   _lateral;  // requested lateral input as a value from -100 to +100
+
+    // custom configs' variables
+    bool    _motor_enabled[UGV_MAX_NUM_MOTORS];
+    float   _throttle_factor[UGV_MAX_NUM_MOTORS];
+    float   _steering_factor[UGV_MAX_NUM_MOTORS];
+    float   _lateral_factor[UGV_MAX_NUM_MOTORS];
+    float   _output_vectored;
+    float   _config_num_of_motors;
```
[the remark this answers is not in the document]
Comment (Ammarf) [reply APMrover2/AP_MotorsUGV.h]: the main purpose of _config_num_of_motors is to avoid errors in pre_arm_check(), otherwise it's not that necessary. I reckon _motor_enabled can also be removed, I'll remove it and do a quick test.

File: APMrover2/AP_MotorsUGV.h [hunk 10]
```diff
@@ -10,6 +10,8 @@ class AP_MotorsUGV {
     // Constructor
     AP_MotorsUGV(AP_ServoRelayEvents &relayEvents);
 
+static const int8_t UGV_MAX_NUM_MOTORS = 6;
```
[the remark this answers is not in the document]
Comment (khancyr) [reply APMrover2/AP_MotorsUGV.h]: humm... I really prefer a static const on this one for two reasons : it is safer, and on generic name like this it is scoped to the class so we won't risk to erase another define somewhere else.

[the remark this answers is not in the document]
Comment (rmackay9) [reply APMrover2/AP_MotorsUGV.h]: @khancyr, OK, if you'd prefer a static const, I'm ok with that.  Where do you think it should be?  Perhaps down in the "protected" section?  above "external references" or "parameters"?

**Comment by rmackay9** [comment]
Here's a link to the discussion with the user of a boat with 4 thrusters. https://discuss.ardupilot.org/t/custom-frame-ardurover/31384/4

**Comment by Ammarf** [comment]
I added OmniX and OmniPlus, I also changed AP_MOTORS_NUM_MOTORS_MAX to a static const, and made the other name changes Randy suggested. 

For the OmniX and OmniPlus, this is the motor order for each one:

roveromni

File: APMrover2/AP_MotorsUGV.cpp [hunk 11]
```diff
@@ -591,6 +614,60 @@ void AP_MotorsUGV::output_skid_steering(bool armed, float steering, float thrott
...
+    if (armed) {
+
+        // clear and set limits based on input
+        set_limits_from_input(armed, steering, throttle);
+
+        // constrain steering
+        steering = constrain_float(steering, -4500.0f, 4500.0f);
+
+        // reverse steering when throttle is negative
+        if (is_negative(throttle)) {
```
Comment (rmackay9) [inline APMrover2/AP_MotorsUGV.cpp]: Although I suspect it's an existing issue with the omni rover code, I don't think we need this reversing of steering when throttle is negative.

It can be tested on a real vehicle by setting [PILOT_STEER_TYPE](http://ardupilot.org/rover/docs/rover-steering-input-type-and-reversing-behaviour.html) = 3 and then putting it into manual mode.  If setup correctly you will find the vehicle will turn in the same direction (i.e. clockwise when right steering input is provided by the user) when throttle is forward or backwards.  I.e. it will rotate in the same direction regardless of whether it is moving forward or backwards.

File: APMrover2/AP_MotorsUGV.cpp [hunk 12]
```diff
@@ -591,6 +614,60 @@ void AP_MotorsUGV::output_skid_steering(bool armed, float steering, float thrott
...
+            thr_str_ltr_out = (scaled_throttle * _throttle_factor[i]) +
+                              (scaled_steering * _steering_factor[i]) +
+                              (scaled_lateral * _lateral_factor[i]);
+            if (fabsf(thr_str_ltr_out) > thr_str_ltr_max) {
+                thr_str_ltr_max = fabsf(thr_str_ltr_out);
+            }
+            float output_vectored = (thr_str_ltr_out / thr_str_ltr_max);
+
+            // send output for each motor
+            SRV_Channels::set_output_scaled(SRV_Channels::get_motor_function(i), 100.0f * output_vectored);
```
Comment (rmackay9) [inline APMrover2/AP_MotorsUGV.cpp]: This change to using set-output-scaled instead of output_throttle() means that users cannot use motors that rely on a "duty cycle" (for speed) and "relay" (for direction) input.  It would be better to stick with using output_throttle() if possible.  I think the relay library supports up to 4 relays so I think it can work.

File: APMrover2/AP_MotorsUGV.cpp (hunk 11 above)
Comment (Ammarf) [reply APMrover2/AP_MotorsUGV.cpp]: You're right, setting PILOT_STEER_TYPE to 3 was enough.

File: APMrover2/AP_MotorsUGV.h [hunk 13]
```diff
@@ -138,5 +157,11 @@ class AP_MotorsUGV {
...
     float   _throttle_prev; // throttle input from previous iteration
     bool    _scale_steering = true; // true if we should scale steering by speed or angle
-    float   _lateral;  // requested lateral input as a value from -4500 to +4500
+    float   _lateral;  // requested lateral input as a value from -100 to +100
+
+    // custom config variables
+    float   _throttle_factor[AP_MOTORS_NUM_MOTORS_MAX];
+    float   _steering_factor[AP_MOTORS_NUM_MOTORS_MAX];
+    float   _lateral_factor[AP_MOTORS_NUM_MOTORS_MAX];
+    float   _motors_num;
```
[the remark this answers is not in the document]
Comment (Ammarf) [reply APMrover2/AP_MotorsUGV.h]: Missed that one. Thanks!

**Comment by rmackay9** [comment]
Merged, thanks!

**Comment by labomat** [comment]
Can this be used to steer a boat with two vectored thrusters (one at thw bow and one at the aft) that can be turned by 270° for steering? I would need two pairs of out put for each thruster (throttle and steering). Is there more documentation about how to use this feature?

**Comment by rmackay9** [comment]
@labomat, no, i'm afraid not.  This particular PR added support for three or four fixed thrusters.  This change is included in Rover-3.5 which has [started beta testing](https://discuss.ardupilot.org/c/ardurover/rover-35).  I hope to add documentation in the coming weeks.

I think we have an issue/enhancement request raised to add support for a bow thruster which is perhaps what you're looking for.  Anything is possible of course, it's just a matter of a developer putting the time in to implement it.

**Comment by labomat** [comment]
A bow thruster would not help, I think. Becaus it is a fixed thruster mounted perpendicular to the main axis - more something like this PR already covered.

**Comment by rmackay9** [comment]
@labomat, ah yes, you're right of course.   I'll reply on the other issue.
